File: built-in plugins dsp/mb-vocal-vocoder.cpp
/**
 * MB Vocoder
 * Category : effect
 * Type     : vocal
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic vocoder synthesis
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_VOCAL_VOCODER_H
#define MB_VOCAL_VOCODER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbVocalVocoder : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-vocal-vocoder";
    static constexpr const char* PLUGIN_NAME    = "MB Vocoder";
    static constexpr const char* PLUGIN_TYPE    = "vocal";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float bands = 16f;  // range [8, 32]
    float formant = 0f;  // range [-12, 12]
    float carrier = 0.5f;  // range [0, 1]
    };

    MbVocalVocoder() = default;
    ~MbVocalVocoder() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.bands = std::clamp(params.bands, 8f, 32f);
        params.formant = std::clamp(params.formant, -12f, 12f);
        params.carrier = std::clamp(params.carrier, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Vocoder
        return input;
    }
};

#endif // MB_VOCAL_VOCODER_H

File: built-in plugins dsp/mb-vocal-vocoder.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbVocalVocoderPlugin: PluginDefinition = { id: 'mb-vocal-vocoder', slug: 'mb-vocal-vocoder', name: 'MB Vocoder', category: 'effect', type: 'vocal', version: '1.0.0', description: 'Classic vocoder synthesis', author: 'Max Booster', grade: 'A', parameters: [{ id: 'bands', name: 'Bands', type: 'float', defaultValue: 16, minValue: 8, maxValue: 32, automatable: false }, { id: 'formant', name: 'Formant', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'carrier', name: 'Carrier Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { bands: 16, formant: 0, carrier: 0.5 } };

export default MbVocalVocoderPlugin;
